sdk/communication/azure-communication-chat/azure/communication/chat/aio/_chat_thread_client_async.py

```diff
@@ -500,14 +501,14 @@ async def add_participants(
         self,
         thread_participants: List[ChatThreadParticipant],
         **kwargs
-    ) -> None:
+    ) -> AddChatParticipantsResult:
```

Comment:
We should unpackage this result type. It's currently pretty hard to use...
If you could write a sample demonstrating what a customer has to do to:
- Add 5 participants
- Determine which 2 our of those 5 failed.
- Re-attempt to add those 2 failures.

From their we can figure out the best response value here.

Reply:
@annatisch something like this?
```python
def _util_create_thread_participant(user, **kwargs) -> ChatThreadParticipant:
    """
    Create util or call the constructor of ChatThreadParticipant - same thing
    """
    return ChatThreadParticipant(
    user=user_id,
    display_name = kwargs.get('display_name', None),
    share_history_time = kwargs.get('share_history_time', None)
    )
	

def decide_if_needs_to_retried(invalid_participant) -> bool:
    """
    some logic to determine if this is needed to be retried or not
    """
    return True


if __name__=='__main__':
    # intialize identity client
    identity_client = CommunicationIdentityClient.from_connection_string(connection_string)
    
    # initialize and retrieve chat_thread_client
    refresh_options = CommunicationTokenRefreshOptions(token)
    chat_client = ChatClient(endpoint, CommunicationTokenCredential(refresh_options))
    topic = "some topic"
    participants = [ChatThreadParticipant(
            user=user, # whoever is creating the chat_thread
            display_name='name',
            share_history_time=datetime.utcnow()
        )]
    create_chat_thread_result = chat_client.create_chat_thread(topic, participants)
    chat_thread_client = chat_client.get_chat_thread_client(create_chat_thread_result.chat_thread.id)
    
    
    # initialize 3+2 participants
    users_for_sucesss = [identity_client.create_user() for i in range(3)] # create 3 users for success - type: List[CommunicationUserIdentifier]
    users_for_failures = [CommunicationUserIdentifier("some random id 1"), CommunicationUserIdentifier("some random id 2")] # create 2 users for failures - type: List[CommunicationUserIdentifier]
    
    # create ChatThreadParticipant objects
    participants_success = [_util_create_thread_participant(user) for user in users_for_sucesss]
    participants_failures = [_util_create_thread_participant(user) for user in users_for_failures]
    
    # add participants to chat thread
    thread_participants = [participants_success + participants_failures]
    add_chat_participants_result = chat_thread_client.add_participants(thread_participants)
    
    # verify if all participants have been added or not
    participants_to_retry = [] # in case of failure store participants to be retried - type: List[CommunicationUserIdentifier]
    
    if hasattr(add_chat_participants_result, 'errors') and \
            add_chat_participants_result.errors is not None:
        print("Error encountered in adding ", len(thread_participants), " participants")
        participant_failed_to_add = add_chat_participants_result.errors.invalid_participants
        
        for failed_participant in participant_failed_to_add:
            # failed_participant.target contains the user_id - type: str
            # For example, in this scenario the values would be either 'some random id 1' or 'some random id 2'
            print("Failed to add participant", failed_participant.target)
            
            if decide_if_needs_to_retried(failed_participant):
                participants_to_retry.append(CommunicationUserIdentifier(failed_participant.target))
    
    # if any user has been failed to be added, then retry
    if len(participants_to_retry) != 0:
        thread_participants = [_util_create_thread_participant(retry_user) for retry_user in participants_to_retry]
        add_chat_participants_result_retry = chat_thread_client.add_participants(thread_participants)
		
```

Reply:
Wow that's lots of code... and a couple of red flags. If a customer ever has to call `hasattr` on the response of call, then we did something wrong :)

If we change the return type to: `List[Tuple[ChatThreadParticipant, Optional[AddParticipantError]]`
Then we could replace all this:
```python
    add_chat_participants_result = chat_thread_client.add_participants(thread_participants)
    
    # verify if all participants have been added or not
    participants_to_retry = [] # in case of failure store participants to be retried - type: List[CommunicationUserIdentifier]
    
    if hasattr(add_chat_participants_result, 'errors') and \
            add_chat_participants_result.errors is not None:
        print("Error encountered in adding ", len(thread_participants), " participants")
        participant_failed_to_add = add_chat_participants_result.errors.invalid_participants
        
        for failed_participant in participant_failed_to_add:
            # failed_participant.target contains the user_id - type: str
            # For example, in this scenario the values would be either 'some random id 1' or 'some random id 2'
            print("Failed to add participant", failed_participant.target)
            
            if decide_if_needs_to_retried(failed_participant):
                participants_to_retry.append(CommunicationUserIdentifier(failed_participant.target))
    
    # if any user has been failed to be added, then retry
    if len(participants_to_retry) != 0:
        thread_participants = [_util_create_thread_participant(retry_user) for retry_user in participants_to_retry]
        add_chat_participants_result_retry = chat_thread_client.add_participants(thread_participants)
```
With just a couple of lines:
```python
result = chat_thread_client.add_participants(thread_participants)
failed = [p for p, error in result if decide_if_needs_to_retried(error)]
retry = chat_thread_client.add_participants(failed)
```

Reply:
@annatisch The following changes have been made
- add_participant - AddChatParticipantsResult -> tuple(ChatThreadParticipant, CommunicationError)
- add_participants - AddChatParticipantsResult -> list(tuple(ChatThreadParticipant, CommunicationError))
- CreateChatThreadResult -> attributes changed to
  - chat_thread -> ChatThread (no change)
  - Errors -> CreateChatThreadErrors -> list(tuple(ChatThreadParticipant, CommunicationError))